scipy/special/xsf/zeta.h

```diff
@@ -1,10 +1,279 @@
+	0.9999999933099243
+    };
+
+    /* Compute riemann_zeta for complex input z using Borwein's algorithm. */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_borwein(std::complex<double> z) {
+	std::complex<double> s1 = 0.0;
+	// Sum in reverse order because smaller terms come later.
+	for (int k = 49; k >= 0; k--) {
+	    double sign = std::pow(-1.0, k);
+	    std::complex<double> den = std::pow(k + 1, z);
+	    auto t1 = sign * (zeta_borwein_coeff[k] - 1.0) / den;
+	    s1 += t1;
+	}
+	std::complex<double> result = s1;
+	result *= -1.0/(1.0 - std::pow(2.0, 1.0 - z));
+	return result;
+    }
+
+    /* Compute riemann zeta for complex z and real part >= 0 */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_right_halfplane(std::complex<double> z) {
+	if (z == 1.0) {
+	    return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	}
+	if (z.real() < 50.0 && std::abs(z.imag()) > 50.0) {
+	    if (std::abs(z.imag()) > 1e9) {
+		/* This is the point where computing with the em series can start to take impractically
+		 * long. Return NaN to avoid hanging indefinitely for large imaginary part. */
+		set_error("zeta", SF_ERROR_NO_RESULT, NULL);
+		return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+	    }
+	    return zeta_euler_maclaurin(z);
+	}
+	return zeta_borwein(z);
+    }
+
+    XSF_HOST_DEVICE inline std::complex<double> exppi(std::complex<double> z) {
+	    // exp(pi*z) for complex z.
+	    double x = z.real();
+	    double y = z.imag();
+	    std::complex<double> factor1(xsf::cospi(y), xsf::sinpi(y));
+	    double factor2 = std::exp(M_PI*x);
+	    return factor1 * factor2;
+	}
+
+    /* Reflection formula for zeta function. 
+     * zeta(z) = 2^z * pi^(z-1) * sin(pi*z/2) * gamma(1 - z) * zeta(1 - z)
+     * Computation is logarithimized to prevent overflow.
+     * TODO: Complexify the cephes zeta_reflection implementation, which uses
+     * the lanczos approximation for the gamma function. */
+    XSF_HOST_DEVICE inline std::complex<double> zeta_reflection(std::complex<double> z) {
+ 	std::complex<double> t1 = z * M_LN2;
+	std::complex<double> t2 = (z - 1.0) * xsf::cephes::detail::LOGPI;
+	std::complex<double> t3;
+	std::complex<double> J(0.0, 1.0);
+	// log(sinpi(z)) using sin(z) = (exp(i*pi*z) - exp(-i*pi*z)) / 2i
+	if (z.imag() > 0 ) {
+	    t3 = -z * J * M_PI / 2.0 + std::log((-1.0 + exppi(z * J)) / (2.0*J));
+	} else {
+	    t3 = z * J * M_PI / 2.0 + std::log((1.0 - exppi(-z * J)) / (2.0*J));
+	}
+	std::complex<double> t4 = xsf::loggamma(1.0 - z);
+	std::complex<double> factor = std::exp(t1 + t2 + t3 + t4);
+	std::complex<double> result = zeta_right_halfplane(1.0 - z);
+	/* zeta tends to 1.0 as real part tends to +inf, so we handle this case
+	 * below to get the correct quadrant of complex infinity for z with real
+	 * part approaching -inf for cases where factor blows up. */
+	return (result == 1.0) ? factor : factor * result;
```

Comment:
I understand that `?` is just a ternary conditional operator, so if `result == 1.0`, `factor * result` is conceptually the same as just `factor` - but is the problem that multiplication by `1.0` _messes up_ the quadrant of the complex infinity? If that's right, please consider rephrasing the comment. (Is `factor` only infinite  if `result` is `1.0`, or is it possible that an infinite factor could be multiplied by a number other than `1.0`? Would that mess it up, too?)`